src/core/text/TextMetrics.js

```diff
@@ -0,0 +1,191 @@
+import Text from './Text';
+
+/**
+ * The TextMetrics object represents the measurement of a block of text with a specified style.
+ *
+ * @class
+ * @memberOf PIXI
+ */
+export default class TextMetrics
+{
+    /**
+     * @param {string} text - the text that was measured
+     * @param {PIXI.TextStyle} style - the style that was measured
+     * @param {number} width - the measured width of the text
+     * @param {number} height - the measured height of the text
+     * @param {array} lines - an array of the lines of text broken by new lines and wrapping if specified in style
+     * @param {array} lineWidths - an array of the line widths for each line matched to `lines`
+     * @param {number} lineHeight - the measured line height for this style
+     * @param {number} maxLineWidth - the maximum line width for all measured lines
+     * @param {Object} fontProperties - the font properties object from Text.calculateFontProperties
+     */
+    constructor(text, style, width, height, lines, lineWidths, lineHeight, maxLineWidth, fontProperties)
+    {
+        this.text = text;
+        this.style = style;
+        this.width = width;
+        this.height = height;
+        this.lines = lines;
+        this.lineWidths = lineWidths;
+        this.lineHeight = lineHeight;
+        this.maxLineWidth = maxLineWidth;
+        this.fontProperties = fontProperties;
+    }
+
+    /**
+     * Measures the supplied string of text and returns a Rectangle.
+     *
+     * @param {string} text - the text to measure.
+     * @param {PIXI.TextStyle} style - the text style to use for measuring
+     * @param {boolean} [wordWrap] - optional override for if word-wrap should be applied to the text.
+     * @param {HTMLCanvasElement} [canvas] - optional specification of the canvas to use for measuring.
+     * @return {PIXI.TextMetrics} measured width and height of the text.
+     */
+    static measure(text, style, wordWrap, canvas)
+    {
+        if (!canvas)
+        {
+            TextMetrics._canvas = TextMetrics._canvas || document.createElement('canvas');
+            canvas = TextMetrics._canvas;
+        }
+
+        wordWrap = wordWrap || style.wordWrap;
+        const font = Text.getFontStyle(style);
+        const fontProperties = Text.calculateFontProperties(font);
+        const context = canvas.getContext('2d');
```

Comment:
Hey @darkfrog26, saw your comment here. 

First off, thanks for getting this PR so far! It's really awesome, and I know it will be useful for devs for doing things like calculating text-fitting. I know the process can feel a little nit-picky sometimes, but with so many people collaborating and using such a big library, we're working hard to keep things consistent, stable and organized. And, this is a pretty sizable refactor of Text, so just wanna make sure we nail it.

How would you feel if we merge this in and then @themoonrat and I can continue to make a few small edits? Not everyone has infinity time to collaborate on PRs, so we are happy to take it off your hands. Let me know.